src/dot_parser/mod.rs

```diff
@@ -0,0 +1,299 @@
+    let mut map = HashMap::new();
+    for a_list in pair.into_inner() {
+        if a_list.as_rule() != Rule::a_list {
+            continue;
+        }
+        let tokens: Vec<_> = a_list.into_inner().collect();
+        let mut i = 0usize;
+        while i < tokens.len() {
+            // tokens are (id) (maybe "=") (maybe id), but pyparsing/pest flattening depends on grammar.
+            // The simple approach here: take token i as key; if token i+1 exists and is not a comma (we filtered commas out in grammar),
+            // treat it as value.
+            let key = tokens[i].as_str().trim().to_string();
+            if i + 1 < tokens.len() {
+                let val = tokens[i + 1].as_str().trim().to_string();
+                map.insert(key, unquote_str(&val));
+                i += 2;
+            } else {
+                map.insert(key, String::new());
+                i += 1;
+            }
+        }
+    }
+    map
+}
+
+/// Extract the first inner token of node_id
+fn node_id_to_string(pair: pest::iterators::Pair<Rule>) -> String {
+    if let Some(child) = pair.into_inner().next() {
+        return unquote_str(child.as_str().trim());
+    }
+    String::new()
+}
+
+#[pyfunction]
+pub fn from_dot(py: Python<'_>, dot_str: &str) -> PyResult<PyObject> {
+    let pairs = DotParser::parse(Rule::graph_file, dot_str).map_err(|e| {
+        PyErr::new::<pyo3::exceptions::PyValueError, _>(format!("DOT parse error: {}", e))
+    })?;
+
+    let mut is_directed = false;
+    let mut node_map: HashMap<String, NodeIndex> = HashMap::new();
+
+    let mut di_graph: Option<PyDiGraph> = None;
+    let mut undi_graph: Option<PyGraph> = None;
+
+    let graph_attrs = PyDict::new(py);
+
+    let mut default_node_attrs: HashMap<String, String> = HashMap::new();
+    let mut default_edge_attrs: HashMap<String, String> = HashMap::new();
+
+    let mut node_attrs_map: HashMap<String, PyObject> = HashMap::new();
+
+    for pair in pairs {
+        if pair.as_rule() != Rule::graph_file {
+            continue;
+        }
+
+        let mut inner = pair.into_inner();
+
+        // handle graph_type
+        let first = inner.next().unwrap();
+        let graph_type_str = if first.as_rule() == Rule::strict {
+            inner.next().unwrap().as_str()
+        } else {
+            first.as_str()
+        };
+        is_directed = graph_type_str == "digraph";
+
+        if is_directed {
+            let graph = StablePyGraph::<Directed>::with_capacity(0, 0);
+            di_graph = Some(PyDiGraph {
+                graph,
+                cycle_state: rustworkx_core::petgraph::algo::DfsSpace::default(),
+                check_cycle: false,
+                node_removed: false,
+                multigraph: true,
+                attrs: graph_attrs.clone().into(),
+            });
+        } else {
+            let graph = StablePyGraph::<Undirected>::with_capacity(0, 0);
+            undi_graph = Some(PyGraph {
```

Comment:
Some thoughts about this handling:
* I'd make a function that is parametrized on StablePyGraph<DirectionTy> e.g. https://github.com/Qiskit/rustworkx/blob/f0240d424f105ada2b0b9af60335094ba69e715b/src/union.rs#L37
* Upon detecting the type, call the parametrization on the if/else

If you need to detect the direction of the graph, call https://docs.rs/petgraph/latest/petgraph/stable_graph/struct.StableGraph.html#method.is_directed within the parametrized function.

I think that choice will simplify your code, especially for handling both `di_graph` and `undi_graph` on every statement